fms/utils/generation.py

```diff
@@ -70,8 +99,23 @@ def generate(
     kwargs["past_key_value_states"] = None
     kwargs["use_cache"] = use_cache
 
-    for _ in range(max_new_tokens):
+    for i in range(max_new_tokens):
         input_ids = next_input[:, -max_seq_len:]
+
+        # compute the attention mask
+        # always need a mask if this is the prompt or if not using a cache
+        if i == 0 or not use_cache:
+            is_pad = input_ids == pad_id
```

Comment:
it seems unsafe to rely on the assumption that a pad ID can always be masked out. in llama, 0=='unk' and there may be unk tokens within a prompt. the presence of unk tokens could be meaningful.

Reply:
I think that is why you can provide your own pad_id param, it just defaults to 0.

Reply:
I guess some models might be trained without masking? though I think all the models we train expect to have some kind of pad/unk/whatever token you can mask out for attention to work as expected when not using nested tensors

Reply:
llama in particular doesn't have any pad token. it was trained without packing, not padding, so every possible pad_id is some specific token in its vocab (like unk or bos or eos)